core/extensions/impl/crypto_extension.cpp

```diff
@@ -188,60 +194,91 @@ namespace kagome::extensions {
 
   // ---------------------- runtime api version 1 methods ----------------------
 
+  namespace {
+    auto encode_result = [](auto &&value) -> common::Buffer {
+      static auto logger =
+          common::createLogger("CryptoExtension::scale_encode");
+      auto &&result = scale::encode(value);
+      if (!result) {
+        logger->error("failed to scale-encode value: {} ",
+                      result.error().message());
+        std::terminate();
+      }
+      return common::Buffer(result.value());
+    };
+
+    template <class T>
+    auto encode_optional = [](auto &&v) -> common::Buffer {
+      return encode_result(std::forward<boost::optional<T>>(v));
+    };
+
+    template <class T, class E>
+    auto encode_variant = [](auto &&v) -> common::Buffer {
+      return encode_result(std::forward<boost::variant<T, E>>(v));
+    };
+  }  // namespace
+
   runtime::WasmSpan CryptoExtension::ext_ed25519_public_keys_v1(
       runtime::WasmSize key_type) {
-    static const common::Buffer error_result{};
+    static const common::Buffer error_result =
+        encode_result(std::vector<crypto::ED25519PublicKey>{});
 
     auto key_type_id = static_cast<crypto::KeyTypeId>(key_type);
     if (!crypto::isSupportedKeyType(key_type_id)) {
       auto kt = crypto::decodeKeyTypeId(key_type_id);
-      logger_->error("key type '{}' is not supported", kt);
-      return memory_->storeBuffer(error_result);
+      logger_->warn("key type '{}' is not officially supported ", kt);
     }
 
     auto &&public_keys = crypto_store_->getEd25519PublicKeys(key_type_id);
-    if (!public_keys) {
-      logger_->error("failed to get ed25519 public keys: {}",
-                     public_keys.error().message());
-      return memory_->storeBuffer(error_result);
-    }
-
-    auto &&encoded = scale::encode(public_keys.value());
-    if (!encoded) {
-      logger_->error("failed to scale-encode vector of public keys: {}",
-                     encoded.error().message());
-      return memory_->storeBuffer(error_result);
-    }
-
-    common::Buffer buffer(std::move(encoded.value()));
+    auto buffer = encode_result(public_keys);
 
     return memory_->storeBuffer(buffer);
   }
 
-  crypto::bip39::Bip39Seed CryptoExtension::deriveBigSeed(
-      std::string_view mnemonic_phrase) {
-    auto &&mnemonic = crypto::bip39::Mnemonic::parse(mnemonic_phrase);
+  common::Blob<32> CryptoExtension::deriveSeed(std::string_view content) {
+    // first check if content is a hexified seed value
+    if (content.substr(0, 2) == "0x") {
+      auto &&res = common::Blob<32>::fromHex(content.substr(2));
+      if (!res) {
+        logger_->error("failed to unhex blob: {}", res.value());
+        std::terminate();
+      }
+    }
+
+    if (auto &&res = common::Blob<32>::fromHex(content); res) {
+      return res.value();
+    }
```

Comment:
If I am not mistaken we already have function in common that unhexes string with prepended 0x. Use it instead

Reply:
That function covers only the case when string has '0x' prefix, otherwise it returns error. So if we use that fucntion, we will still consider 2 cases: with and without prefix. And after that we will need to copy value to blob. It doesn't look more simple. Here I assume that if value starts with '0x' prefix, it means a hexified seed, otherwise it still can be hexified value or mnemonic, thus if unhexing fails, we try decode mnemonic.

Reply:
We should not have two cases of hex strings: with and without 0x in keystore. Assume there is only one case: with 0x prefixed. Hex value without 0x can be considered invalid.

Reply:
refactored